pkg/controller/direct/kms/autokeyconfig_controller.go

```diff
@@ -0,0 +1,273 @@
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	ctrlName      = "kms-controller"
+	serviceDomain = "//cloudkms.googleapis.com"
+)
+
+func init() {
+	registry.RegisterModel(krm.KMSAutokeyConfigGVK, NewModel)
+}
+
+func NewModel(ctx context.Context, config *config.ControllerConfig) (directbase.Model, error) {
+	return &model{config: *config}, nil
+}
+
+var _ directbase.Model = &model{}
+
+type model struct {
+	config config.ControllerConfig
+}
+
+func (m *model) client(ctx context.Context) (*gcp.AutokeyAdminClient, error) {
+	var opts []option.ClientOption
+	opts, err := m.config.RESTClientOptions()
+	if err != nil {
+		return nil, err
+	}
+	gcpClient, err := gcp.NewAutokeyAdminRESTClient(ctx, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("building AutokeyConfig client: %w", err)
+	}
+	return gcpClient, err
+}
+
+func (m *model) AdapterForObject(ctx context.Context, reader client.Reader, u *unstructured.Unstructured) (directbase.Adapter, error) {
+	obj := &krm.KMSAutokeyConfig{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	var folder *refs.Folder
+	var err error
+	folder, err = refs.ResolveFolderFromAnnotation(ctx, reader, obj)
+	if err != nil {
+		folder, err = refs.ResolveFolder(ctx, reader, obj, obj.Spec.FolderRef)
+	}
+	if err != nil || folder.FolderID == "" {
+		return nil, fmt.Errorf("unable to resolve folder for autokeyConfig name: %s", obj.GetName())
+	}
+	var id *KMSAutokeyConfigIdentity
+	externalRef := direct.ValueOf(obj.Status.ExternalRef)
+	if externalRef == "" {
+		id = BuildID(folder.FolderID)
+	} else {
+		id, err = asID(externalRef)
+		if err != nil {
+			return nil, err
+		}
+
+		if id.Parent.FolderID != folder.FolderID {
+			return nil, fmt.Errorf("KMSAutokeyConfig %s/%s has spec.folderRef changed, expect %s, got %s",
+				u.GetNamespace(), u.GetName(), id.Parent.FolderID, folder.FolderID)
+		}
+	}
+
+	gcpClient, err := m.client(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &Adapter{
+		id:        id,
+		gcpClient: gcpClient,
+		desired:   obj,
+	}, nil
+}
+
+func (m *model) AdapterForURL(ctx context.Context, url string) (directbase.Adapter, error) {
+	// TODO: Support URLs
+	return nil, nil
+}
+
+type Adapter struct {
+	id        *KMSAutokeyConfigIdentity
+	gcpClient *gcp.AutokeyAdminClient
+	desired   *krm.KMSAutokeyConfig
+	actual    *kmspb.AutokeyConfig
+}
+
+var _ directbase.Adapter = &Adapter{}
+
+func (a *Adapter) Find(ctx context.Context) (bool, error) {
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("getting KMSAutokeyConfig", "name", a.id.FullyQualifiedName())
+	req := &kmspb.GetAutokeyConfigRequest{Name: a.id.FullyQualifiedName()}
+	autokeyconfigpb, err := a.gcpClient.GetAutokeyConfig(ctx, req)
```

Comment:
Does KMS returns 404 if a autokeyConfig is not found? From the offline discussion, I was in the impression that it will get the object but with `keyProject` set to `uninitialized`, right?

Reply:
there will never be 404.
I have gone through the api code and it seems, 
1. if the folder exist and the user has permissions to access autokeyConfig then we will return valid response. 
The status will be unitialized if the key_project empty.
2. else, permission denied exception will be thrown by the api